File: C++/Pattern wise question/BFS/4_average_of_level.cpp
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> res;
        if(root == NULL)
            return res;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int sz = q.size();
            double sum=0;
            for(int i=0; i<sz; i++){
                 TreeNode* temp = q.front();
                 q.pop();
                 sum += temp->val;

                if(temp->left) q.push(temp->left);

                if(temp->right) q.push(temp->right);

            }
            res.push_back(sum / sz);
        }
        return res;
    }
};

File: Python/Pattern wise question/BFS/4_average_of_level.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        res = []
        if(root == None):
            return res
        q = deque()
        q.append(root)

        while(q):
            sz = len(q)
            add = 0
            for i in range(sz):
                temp = q.popleft()
                add += temp.val

                if(temp.left):
                    q.append(temp.left)
                if(temp.right):
                    q.append(temp.right)

            res.append(add/sz)

        return res
        
